engine/src/main/java/org/terasology/logic/players/FirstPersonClientSystem.java

```diff
@@ -131,50 +138,61 @@ public void onHeldItemChanged(OnChangedComponent event, EntityRef character, Cha
         if (localPlayer.getCharacterEntity().equals(character)) {
             EntityRef oldHeldItem = currentHeldItem;
             currentHeldItem = heldItemComponent.selectedItem;
-            linkHeldItemLocationForLocalPlayer(character, currentHeldItem, oldHeldItem);
-        }
-    }
-
-    @ReceiveEvent(netFilter = RegisterMode.CLIENT)
-    public void onLocationRemovedAddClientSideLocation(BeforeDeactivateComponent event, EntityRef entityRef, LocationComponent locationComponent) {
-        // when in a remote client situation,  the remove LocationComponent happens after the CharacterHeldItemComponent is changed.
-        // So this allows re-adding the client side LocationComponent after this
-        if (entityRef.equals(currentHeldItem)) {
-            relinkHeldItem = true;
+            linkHeldItemLocationForLocalPlayer(currentHeldItem, oldHeldItem);
         }
     }
 
-    void linkHeldItemLocationForLocalPlayer(EntityRef character, EntityRef newItem, EntityRef oldItem) {
+    /**
+     * Changes held item entity.
+     *
+     * <p>Detaches old held item and removes it's components. Adds components to new held item and
+     * attaches it to the mount point entity.</p>
+     */
+    private void linkHeldItemLocationForLocalPlayer(EntityRef newItem, EntityRef oldItem) {
         if (!newItem.equals(oldItem)) {
             EntityRef camera = localPlayer.getCameraEntity();
             FirstPersonHeldItemMountPointComponent mountPointComponent = camera.getComponent(FirstPersonHeldItemMountPointComponent.class);
             if (mountPointComponent != null) {
+
+                if (clientHeldItem.exists()) {
+                    clientHeldItem.destroy();
+                    clientHeldItem = EntityRef.NULL;
+                }
+
+
                 // remove the location from the old item
                 if (oldItem != null && oldItem.exists()) {
-                    Location.removeChild(mountPointComponent.mountPointEntity, oldItem);
-                    oldItem.removeComponent(LocationComponent.class);
                     oldItem.removeComponent(ItemIsHeldComponent.class);
                 } else {
-                    Location.removeChild(mountPointComponent.mountPointEntity, getHandEntity());
-                    getHandEntity().removeComponent(LocationComponent.class);
                     getHandEntity().removeComponent(ItemIsHeldComponent.class);
                 }
 
                 // use the hand if there is no new item
-                EntityRef heldItem = newItem;
-                if (!heldItem.exists()) {
+                EntityRef heldItem;
+                if (!newItem.exists()) {
                     heldItem = getHandEntity();
+                } else {
+                    heldItem = newItem;
```

Comment:
agh, it didn't select the amount of lines I tohught it would. Changing the few lines after the comment `// use the hand if there is no new item` seems to just do the same thing as before, just structured in a different way.